crates/hc/src/lib.rs

```diff
@@ -108,113 +108,33 @@
 //! ```
 //! and the examples.
 
-use std::path::Path;
-use std::path::PathBuf;
-
-use holochain_conductor_api::{AdminRequest, AdminResponse};
-use holochain_websocket::WebsocketSender;
-use ports::get_admin_api;
-
-pub use ports::force_admin_port;
-
-/// Print a msg with `hc-admin: ` pre-pended
-/// and ansi colors.
-macro_rules! msg {
-    ($($arg:tt)*) => ({
-        use ansi_term::Color::*;
-        print!("{} ", Blue.bold().paint("hc-admin:"));
-        println!($($arg)*);
-    })
-}
-
-pub mod calls;
-#[doc(hidden)]
-pub mod cmds;
-pub mod config;
-pub mod dna;
-pub mod generate;
-pub mod run;
-pub mod save;
-pub mod setups;
-
-mod ports;
-
-/// An active connection to a running conductor.
-pub struct CmdRunner {
-    client: WebsocketSender,
+use holochain_cli_bundle as hc_bundle;
+use holochain_cli_sandbox as hc_sandbox;
+use structopt::StructOpt;
+
+/// Holochain CLI
+///
+/// Work with DNA and hApp bundle files, set up sandbox environments for testing
+/// and development purposes, make direct admin calls to running conductors,
+/// and more.
+#[derive(Debug, StructOpt)]
+pub enum Opt {
+    /// Work with hApp bundles
+    App(hc_bundle::HcAppBundle),
+    /// Work with DNA bundles
+    Dna(hc_bundle::HcDnaBundle),
+    /// Work with sandboxed environments for testing and development
+    Sandbox(hc_sandbox::HcSandbox),
```

Comment:
I'm a little confused.

- On the one hand, we seem to have a monolithic `hc` binary here which includes all the code for packaging and sandboxing.
- We also have individual bin files for hc-app, hc-dna, and hc-sandbox.

Should we not pick one? I.e. we can not depend directly on the sub-crates, but delegate to their binary capabilities if found on the path like cargo?

Reply:
Yeah, so the monolithic `hc` is the intended use. I added bin builds for the subcrates because "why not?". I wouldn't expect anyone to actually build and use those.

Can you describe in more detail how cargo does it? Are you saying that `cargo` is a monolith but its subcommands can be overridden by the presence of `cargo-subcommand` on the PATH?

Reply:
No cargo subcommand logic is built into the base cargo executable. When you invoke a sub-command cargo looks for `cargo-[subcommand]` on the path, if it finds it, it runs it, forwarding all cli parameters.

So the top-level `cargo` binary is agnostic to its own sub-commands.